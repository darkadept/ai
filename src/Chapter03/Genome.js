import slice from 'lodash/slice';
import {random} from '../lib';

export default class Genome {
	constructor(numBits) {
		this._vecBits = [];
		this._fitness = 0;

		for (let i=0; i<numBits; i++) {
			this._vecBits.push(random(0, 1));
		}
	}

	get fitness() {return this._fitness};
	set fitness(value) {this._fitness = value;}

	get length() {return this._vecBits.length};
	slice(start, numBits) {
		return slice(this._vecBits, start, start+numBits);
	}
	// get bits() {return this._vecBits;}

	// equals(b) {
	// 	return this._vecBits.join('') === b._vecBits.join('');
	// }
}
