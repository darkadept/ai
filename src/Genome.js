import {random} from './lib';

export default class Genome {
	constructor(numBits) {
		this._vecBits = [];
		this._fitness = 0;

		for (let i=0; i<numBits; i++) {
			this._vecBits.push(random(0, 1));
		}
	}

	get fitness() {return this._fitness};
	set fitness(value) {this._fitness = value;}
	get vector() {return this._vecBits;}
}
