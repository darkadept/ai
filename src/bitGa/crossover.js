import {randomFloat, random} from '../lib';
import Genome from './Genome';

/**
 * Crossover bits of a mum and dad genome. Returns two babies.
 * @param mum
 * @param dad
 * @param crossoverRate
 * @returns {*}
 */
export default function crossover({mum, dad, crossoverRate}) {
	if (mum === dad || randomFloat() > crossoverRate) {
		return {
			baby1: mum,
			baby2: dad,
		};
	}

	const cp = random(0, mum.length - 1);

	let baby1 = new Genome({bits: [
			...mum.slice(0, cp),
			...dad.slice(cp, dad.length - cp),
	]});
	let baby2 = new Genome({
		bits: [
			...dad.slice(0, cp),
			...mum.slice(cp, mum.length - cp),
		]
	});

	return {baby1, baby2};
}
