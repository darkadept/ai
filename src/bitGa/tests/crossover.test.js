import Genome from '../Genome';
import crossover from '../crossover';

const mockMath = Object.create(global.Math);
let randomFlip = true;
mockMath.random = () => {
	if (randomFlip) {
		randomFlip = false;
		return 0.3;
	}
	randomFlip = true;
	return 0.6;
};
global.Math = mockMath;

it('should cross over bits', () => {
	// const mum = new Genome({bits: [0, 1, 0, 1, 0, 0, 1, 1]});
	// const dad = new Genome({bits: [1, 1, 1, 0, 1, 0, 1, 0]});
	const mum = new Genome({bits: [0, 0, 0, 0, 1, 1, 1, 1]});
	const dad = new Genome({bits: [1, 1, 1, 1, 0, 0, 0, 0]});

	expect(mum.bits).toMatchSnapshot();
	expect(dad.bits).toMatchSnapshot();

	const babies = crossover({mum, dad, crossoverRate: 0.8});
	expect(babies).toMatchSnapshot();
});
