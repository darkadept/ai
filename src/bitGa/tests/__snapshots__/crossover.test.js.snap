// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should cross over bits 1`] = `
Array [
  0,
  0,
  0,
  0,
  1,
  1,
  1,
  1,
]
`;

exports[`should cross over bits 2`] = `
Array [
  1,
  1,
  1,
  1,
  0,
  0,
  0,
  0,
]
`;

exports[`should cross over bits 3`] = `
Object {
  "baby1": Genome {
    "_bits": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "_fitness": 0,
  },
  "baby2": Genome {
    "_bits": Array [
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
    ],
    "_fitness": 0,
  },
}
`;
